File: apps/desktop/src/lib/project-status.ts
import { invoke as tauriInvoke } from '@tauri-apps/api/core';
import { CURRENT_PROTOCOL_VERSION, type InvokeCommand } from './health';
import { parseProjectSummaryV1, type ProjectSummaryV1 } from './project';

const STABLE_ID_PATTERN = /^[0-9a-f]{64}$/;
const GENERATION_PATTERN = /^[1-9][0-9]{0,18}$/;
const MAX_GENERATION = 9_223_372_036_854_775_807n;
const BYTE_COUNT_PATTERN = /^(?:0|[1-9][0-9]{0,19})$/;
const MAX_BYTE_COUNT = 18_446_744_073_709_551_615n;

export interface QueryProjectStatusRequestV1 {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
}

export type IndexStateV1 = 'notStarted' | 'building' | 'published' | 'failed' | 'cancelled';
export type RebuildStateV1 = 'idle' | 'queued' | 'running' | 'succeeded' | 'failed' | 'cancelled';

export interface ProjectSnapshotV1 {
  generation: string;
  snapshotId: string;
}

export interface ProjectIndexStatusV1 {
  latestAttemptSnapshotId: string | null;
  latestSnapshot: ProjectSnapshotV1 | null;
  publishedSnapshotId: string | null;
  state: IndexStateV1;
}

export type ProjectStatusResultV1 =
  | { status: 'noProject' }
  | {
      index: ProjectIndexStatusV1;
      project: ProjectSummaryV1;
      projectId: string;
      rebuildState: RebuildStateV1;
      status: 'active';
      storageBytes: string | null;
    };

export interface ProjectStatusResponseV1 {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
  result: ProjectStatusResultV1;
}

const invokeThroughTauri: InvokeCommand = (command, arguments_) =>
  tauriInvoke<unknown>(command, arguments_);

export async function queryProjectStatus(
  invokeCommand: InvokeCommand = invokeThroughTauri,
): Promise<ProjectStatusResponseV1> {
  const request: QueryProjectStatusRequestV1 = { protocolVersion: CURRENT_PROTOCOL_VERSION };
  const payload = await invokeCommand('query_project_status', { request });
  return parseProjectStatusResponseV1(payload);
}

export function parseProjectStatusResponseV1(payload: unknown): ProjectStatusResponseV1 {
  if (
    !isRecord(payload) ||
    !hasExactKeys(payload, ['protocolVersion', 'result']) ||
    payload.protocolVersion !== CURRENT_PROTOCOL_VERSION
  ) {
    throw new Error('Project status response does not match the V1 schema.');
  }

  return {
    protocolVersion: payload.protocolVersion,
    result: parseResult(payload.result),
  };
}

function parseResult(value: unknown): ProjectStatusResultV1 {
  if (!isRecord(value) || typeof value.status !== 'string') {
    throw new Error('Project status response contains an invalid result.');
  }
  if (value.status === 'noProject' && hasExactKeys(value, ['status'])) {
    return { status: 'noProject' };
  }
  if (
    value.status === 'active' &&
    hasExactKeys(value, [
      'index',
      'project',
      'projectId',
      'rebuildState',
      'status',
      'storageBytes',
    ]) &&
    isStableId(value.projectId) &&
    isStorageBytes(value.storageBytes) &&
    isRebuildState(value.rebuildState)
  ) {
    return {
      index: parseIndexStatus(value.index),
      project: parseProjectSummaryV1(value.project),
      projectId: value.projectId,
      rebuildState: value.rebuildState,
      status: 'active',
      storageBytes: value.storageBytes,
    };
  }
  throw new Error('Project status response contains an invalid result.');
}

function parseIndexStatus(value: unknown): ProjectIndexStatusV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, [
      'latestAttemptSnapshotId',
      'latestSnapshot',
      'publishedSnapshotId',
      'state',
    ]) ||
    !isIndexState(value.state) ||
    !isOptionalStableId(value.latestAttemptSnapshotId) ||
    !isOptionalStableId(value.publishedSnapshotId)
  ) {
    throw new Error('Project status response contains an invalid index state.');
  }

  const latestSnapshot = value.latestSnapshot === null ? null : parseSnapshot(value.latestSnapshot);
  const notStarted = value.state === 'notStarted';
  if (
    notStarted !== (value.latestAttemptSnapshotId === null) ||
    (value.latestAttemptSnapshotId !== null && latestSnapshot === null) ||
    (value.publishedSnapshotId !== null && latestSnapshot === null)
  ) {
    throw new Error('Project status response contains inconsistent index metadata.');
  }

  return {
    latestAttemptSnapshotId: value.latestAttemptSnapshotId,
    latestSnapshot,
    publishedSnapshotId: value.publishedSnapshotId,
    state: value.state,
  };
}

function parseSnapshot(value: unknown): ProjectSnapshotV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, ['generation', 'snapshotId']) ||
    !isStableId(value.snapshotId) ||
    typeof value.generation !== 'string' ||
    !GENERATION_PATTERN.test(value.generation) ||
    BigInt(value.generation) > MAX_GENERATION
  ) {
    throw new Error('Project status response contains an invalid snapshot.');
  }
  return { generation: value.generation, snapshotId: value.snapshotId };
}

function isIndexState(value: unknown): value is IndexStateV1 {
  return (
    value === 'notStarted' ||
    value === 'building' ||
    value === 'published' ||
    value === 'failed' ||
    value === 'cancelled'
  );
}

function isRebuildState(value: unknown): value is RebuildStateV1 {
  return (
    value === 'idle' ||
    value === 'queued' ||
    value === 'running' ||
    value === 'succeeded' ||
    value === 'failed' ||
    value === 'cancelled'
  );
}

function isStableId(value: unknown): value is string {
  return typeof value === 'string' && STABLE_ID_PATTERN.test(value);
}

function isOptionalStableId(value: unknown): value is string | null {
  return value === null || isStableId(value);
}

function isStorageBytes(value: unknown): value is string | null {
  return (
    value === null ||
    (typeof value === 'string' && BYTE_COUNT_PATTERN.test(value) && BigInt(value) <= MAX_BYTE_COUNT)
  );
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function hasExactKeys(value: Record<string, unknown>, expected: string[]): boolean {
  const keys = Object.keys(value).sort();
  const sortedExpected = [...expected].sort();
  return (
    keys.length === sortedExpected.length &&
    keys.every((key, index) => key === sortedExpected[index])
  );
}

File: crates/a3-protocol/src/project_status.rs
use crate::{ProjectSummaryV1, ProtocolVersion};
use serde::{Deserialize, Serialize};

/// Strict input payload for the V1 active-project status query.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct QueryProjectStatusRequestV1 {
    protocol_version: ProtocolVersion,
}

impl QueryProjectStatusRequestV1 {
    /// Creates a request for a specific protocol version.
    #[must_use]
    pub const fn new(protocol_version: ProtocolVersion) -> Self {
        Self { protocol_version }
    }

    /// Creates a request for the protocol version emitted by this build.
    #[must_use]
    pub const fn current() -> Self {
        Self::new(ProtocolVersion::CURRENT)
    }

    /// Returns the requested protocol version.
    #[must_use]
    pub const fn protocol_version(self) -> ProtocolVersion {
        self.protocol_version
    }
}

/// Versioned response describing the Core-owned active project, if present.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ProjectStatusResponseV1 {
    protocol_version: ProtocolVersion,
    result: ProjectStatusResultV1,
}

impl ProjectStatusResponseV1 {
    /// Creates the response used before any project has been selected.
    #[must_use]
    pub const fn no_project() -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result: ProjectStatusResultV1::NoProject,
        }
    }

    /// Creates a bounded status response for the validated Core-owned project.
    #[must_use]
    pub fn active(
        project_id: String,
        project: ProjectSummaryV1,
        index: ProjectIndexStatusV1,
        storage_bytes: Option<String>,
        rebuild_state: RebuildStateV1,
    ) -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result: ProjectStatusResultV1::Active {
                project_id,
                project: Box::new(project),
                index: Box::new(index),
                storage_bytes,
                rebuild_state,
            },
        }
    }

    /// Returns the response protocol version.
    #[must_use]
    pub const fn protocol_version(&self) -> ProtocolVersion {
        self.protocol_version
    }

    /// Returns the mutually exclusive active-project result.
    #[must_use]
    pub const fn result(&self) -> &ProjectStatusResultV1 {
        &self.result
    }
}

/// Active-project result selected entirely by privileged Core state.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase", tag = "status")]
pub enum ProjectStatusResultV1 {
    /// No project has been opened in this desktop process.
    NoProject,
    /// A validated project is active and its bounded index metadata was loaded.
    Active {
        /// Stable catalog identity assigned by local storage.
        #[serde(rename = "projectId")]
        project_id: String,
        /// Existing WebView-safe repository and worktree projection.
        project: Box<ProjectSummaryV1>,
        /// Durable snapshot and index-run projection.
        index: Box<ProjectIndexStatusV1>,
        /// Exact private A^3 storage usage as lossless decimal text.
        #[serde(rename = "storageBytes")]
        storage_bytes: Option<String>,
        /// Lifecycle of the latest user-requested regenerable-index rebuild.
        #[serde(rename = "rebuildState")]
        rebuild_state: RebuildStateV1,
    },
}

/// WebView-safe durable index status for one worktree.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ProjectIndexStatusV1 {
    state: IndexStateV1,
    latest_snapshot: Option<ProjectSnapshotV1>,
    latest_attempt_snapshot_id: Option<String>,
    published_snapshot_id: Option<String>,
}

impl ProjectIndexStatusV1 {
    /// Creates a status from already validated application projections.
    #[must_use]
    pub const fn new(
        state: IndexStateV1,
        latest_snapshot: Option<ProjectSnapshotV1>,
        latest_attempt_snapshot_id: Option<String>,
        published_snapshot_id: Option<String>,
    ) -> Self {
        Self {
            state,
            latest_snapshot,
            latest_attempt_snapshot_id,
            published_snapshot_id,
        }
    }

    /// Returns the latest durable attempt state or `notStarted`.
    #[must_use]
    pub const fn state(&self) -> IndexStateV1 {
        self.state
    }

    /// Returns the latest observed immutable snapshot.
    #[must_use]
    pub const fn latest_snapshot(&self) -> Option<&ProjectSnapshotV1> {
        self.latest_snapshot.as_ref()
    }

    /// Returns the input snapshot of the latest index attempt.
    #[must_use]
    pub fn latest_attempt_snapshot_id(&self) -> Option<&str> {
        self.latest_attempt_snapshot_id.as_deref()
    }

    /// Returns the snapshot still visible through the publish boundary.
    #[must_use]
    pub fn published_snapshot_id(&self) -> Option<&str> {
        self.published_snapshot_id.as_deref()
    }
}

/// Durable lifecycle projection of the latest index attempt.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum IndexStateV1 {
    /// No index attempt has been recorded.
    NotStarted,
    /// The latest attempt currently owns the index mutation slot.
    Building,
    /// The latest attempt was atomically published.
    Published,
    /// The latest attempt failed without replacing a previous publication.
    Failed,
    /// The latest attempt was cancelled without replacing a previous publication.
    Cancelled,
}

/// Core-owned lifecycle of a user-requested regenerable-index rebuild.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum RebuildStateV1 {
    /// No rebuild has been requested for the active project.
    Idle,
    /// The coordinator accepted the request and is quiescing prior index work.
    Queued,
    /// The bounded rebuild job is deleting only regenerable projections.
    Running,
    /// Regenerable projections were removed and an authoritative refresh was requested.
    Succeeded,
    /// The rebuild failed without deleting authoritative task or snapshot history.
    Failed,
    /// The owning scheduler cancelled the rebuild before commit.
    Cancelled,
}

/// Bounded identity and generation of the latest durable snapshot.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ProjectSnapshotV1 {
    snapshot_id: String,
    generation: String,
}

impl ProjectSnapshotV1 {
    /// Creates a snapshot projection using a decimal string for lossless JavaScript transport.
    #[must_use]
    pub const fn new(snapshot_id: String, generation: String) -> Self {
        Self {
            snapshot_id,
            generation,
        }
    }

    /// Returns the lowercase snapshot digest.
    #[must_use]
    pub fn snapshot_id(&self) -> &str {
        &self.snapshot_id
    }

    /// Returns the positive worktree generation as canonical decimal text.
    #[must_use]
    pub fn generation(&self) -> &str {
        &self.generation
    }
}

#[cfg(test)]
mod tests {
    use super::{
        IndexStateV1, ProjectIndexStatusV1, ProjectSnapshotV1, ProjectStatusResponseV1,
        RebuildStateV1,
    };
    use crate::{GitHeadV1, ProjectSummaryV1};
    use serde_json::json;

    #[test]
    fn active_project_status_has_a_stable_bounded_shape() -> Result<(), serde_json::Error> {
        let response = ProjectStatusResponseV1::active(
            "33".repeat(32),
            ProjectSummaryV1::new(
                "11".repeat(32),
                "22".repeat(32),
                "/worktree".to_owned(),
                GitHeadV1::Unborn {
                    reference: "refs/heads/main".to_owned(),
                },
            ),
            ProjectIndexStatusV1::new(
                IndexStateV1::Published,
                Some(ProjectSnapshotV1::new("44".repeat(32), "7".to_owned())),
                Some("44".repeat(32)),
                Some("44".repeat(32)),
            ),
            Some("4096".to_owned()),
            RebuildStateV1::Idle,
        );

        assert_eq!(
            serde_json::to_value(response)?,
            json!({
                "protocolVersion": 1,
                "result": {
                    "status": "active",
                    "projectId": "33".repeat(32),
                    "project": {
                        "repositoryId": "11".repeat(32),
                        "worktreeId": "22".repeat(32),
                        "worktreeRootDisplay": "/worktree",
                        "head": { "kind": "unborn", "reference": "refs/heads/main" }
                    },
                    "index": {
                        "state": "published",
                        "latestSnapshot": {
                            "snapshotId": "44".repeat(32),
                            "generation": "7"
                        },
                        "latestAttemptSnapshotId": "44".repeat(32),
                        "publishedSnapshotId": "44".repeat(32)
                    },
                    "storageBytes": "4096",
                    "rebuildState": "idle"
                }
            })
        );
        Ok(())
    }
}
